### PR TITLE
Add metadata to v2 Org VDC networks

### PR DESCRIPTION
**Pull request by adambarreiro**
Created by #372

Depends on https://github.com/vmware/go-vcloud-director/pull/459

This PR adds `metadata` support for the **v2** VDC networks.

### CODE REVIEW

File: .changes/v3.6.0/816-improvements.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+* `resource/vcd_network_isolated_v2` add support for `metadata` so this field can be set when creating/updating isolated networks. [GH-816]
+* `datasource/vcd_network_isolated_v2` add support for `metadata` so this field can be retrieved when reading from isolated networks. [GH-816]
+* `resource/vcd_network_routed_v2` add support for `metadata` so this field can be set when creating/updating routed networks. [GH-816]
+* `datasource/vcd_network_routed_v2` add support for `metadata` so this field can be retrieved when reading from routed networks. [GH-816]
```
Comment (lvirbalas) [inline .changes/v3.6.0/816-improvements.md:4]: Maybe we should add note that using XML API until OpenAPI support is in alpha? Or are you expecting that nothing will have to be changed in Terraform - that we'll just upgrade the govcd?

File: .changes/v3.6.0/816-improvements.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+* `resource/vcd_network_isolated_v2` add support for `metadata` so this field can be set when creating/updating isolated networks. [GH-816]
```
Comment (lvirbalas) [inline .changes/v3.6.0/816-improvements.md:1]: Looks like docs for the new fields are missing?

**Review by lvirbalas — CHANGES_REQUESTED** [review]

File: .changes/v3.6.0/816-improvements.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
+* `datasource/vcd_network_isolated_v2` add support for `metadata` so this field can be retrieved when reading from isolated networks. [GH-816]
```
Comment (lvirbalas) [inline .changes/v3.6.0/816-improvements.md:2]: I'm just thinking about the casual readers of the changelog. Maybe we should add "NSX-T" in all these entries, so people know the networking type unambiguously? Something like:

```suggestion
* `datasource/vcd_network_isolated_v2` add support for `metadata` so this field can be retrieved when reading from isolated NSX-T networks. [GH-816]
```

File: .changes/v3.6.0/816-improvements.md (hunk 2 above)
Comment (adambarreiro) [reply .changes/v3.6.0/816-improvements.md]: Thanks for noticing, I need to be more careful, I also forgot the entries of #804. I've added them all.

**Review by lvirbalas — APPROVED** [review]

File: .changes/v3.6.0/816-improvements.md (hunk 3 above)
Comment (adambarreiro) [reply .changes/v3.6.0/816-improvements.md]: Done. Not sure if we should put instead _"NSX-T and NSX-V"_, as if I'm not wrong the **v2** resource supports both. But probably just noting NSX-T is clearer for the user.

File: vcd/resource_vcd_network_routed_v2.go [hunk 4]
```diff
@@ -211,10 +227,20 @@ func resourceVcdNetworkRoutedV2Read(ctx context.Context, d *schema.ResourceData,
 
 	d.SetId(orgNetwork.OpenApiOrgVdcNetwork.ID)
 
+	metadata, err := orgNetwork.GetMetadata()
+	if err != nil {
+		log.Printf("[DEBUG] Unable to find routed network v2 metadata: %s", err)
+		return diag.Errorf("[routed network read v2] unable to find Org VDC network metadata %s", err)
```
Comment (vbauzysvmware) [inline vcd/resource_vcd_network_routed_v2.go:233]:
```suggestion
		return diag.Errorf("[routed network read v2] unable to find Org VDC routed network metadata %s", err)
```
Let's be more specific in all places

File: .changes/v3.6.0/816-improvements.md [hunk 5]
```diff
@@ -0,0 +1,4 @@
+* `resource/vcd_network_isolated_v2` add support for `metadata` so this field can be set when creating/updating isolated NSX-T networks. [GH-816]
```
Comment (vbauzysvmware) [inline .changes/v3.6.0/816-improvements.md:1]: direct network details missing

**Review by vbauzysvmware — CHANGES_REQUESTED** [review]

File: .changes/v3.6.0/816-improvements.md (hunk 5 above)
Comment (adambarreiro) [reply .changes/v3.6.0/816-improvements.md:1]: Done in #804 (for _v1_ networks), as direct networks don't have _v2_

File: vcd/datasource_vcd_network_isolated_v2_nsxt_test.go [hunk 6]
```diff
@@ -20,10 +20,14 @@ func TestAccVcdNetworkIsolatedV2NsxtDS(t *testing.T) {
...
+		"NsxtVdc":              testConfig.Nsxt.Vdc,
+		"NetworkName":          t.Name(),
+		"Tags":                 "network nsxt",
+		"MetadataKey":          "key1",
+		"MetadataValue":        "value1",
+		"MetadataKeyUpdated":   "key2",
+		"MetadataValueUpdated": "value2",
 	}
 
 	params["FuncName"] = t.Name() + "-DS"
```
[the remark this answers is not in the document]
Comment (adambarreiro) [reply vcd/datasource_vcd_network_isolated_v2_nsxt_test.go:33]: In this case I didn't need to add anything but the params. Reason is, datasource test is using same template code as resource test `resource_vcd_network_isolated_v2_nsxt_test.go` (which is `testAccVcdNetworkIsolatedV2NsxtStep1`). Then, the test was already comparing that the resource fields are exactly the same as the datasource, hence I didn't need to change it.

File: vcd/datasource_vcd_network_routed_v2_nsxt_test.go [hunk 7]
```diff
@@ -20,11 +20,15 @@ func TestAccVcdNetworkRoutedV2NsxtDS(t *testing.T) {
...
+		"Org":                  testConfig.VCD.Org,
+		"NsxtVdc":              testConfig.Nsxt.Vdc,
+		"EdgeGw":               testConfig.Nsxt.EdgeGateway,
+		"NetworkName":          t.Name(),
+		"Tags":                 "network nsxt",
+		"MetadataKey":          "key1",
+		"MetadataValue":        "value1",
+		"MetadataKeyUpdated":   "key2",
+		"MetadataValueUpdated": "value2",
 	}
```
[the remark this answers is not in the document]
Comment (adambarreiro) [reply vcd/datasource_vcd_network_routed_v2_nsxt_test.go:33]: Same as above, test is pretty similar.

File: vcd/datasource_vcd_network_isolated_v2_nsxt_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/datasource_vcd_network_isolated_v2_nsxt_test.go:33]: great

File: .changes/v3.6.0/816-improvements.md (hunk 5 above)
Comment (vbauzysvmware) [reply .changes/v3.6.0/816-improvements.md:1]: ah, ok, only documentation was updated

File: vcd/resource_vcd_network_isolated_v2.go [hunk 8]
```diff
@@ -167,10 +178,15 @@ func resourceVcdNetworkIsolatedV2Update(ctx context.Context, d *schema.ResourceD
 		return diag.Errorf("[isolated network v2 update] error updating Org VDC network: %s", err)
 	}
 
+	err = createOrUpdateOpenApiNetworkMetadata(d, orgNetwork)
+	if err != nil {
+		return diag.Errorf("[isolated network v2 update] error updating Org VDC network metadata: %s", err)
```
[the remark this answers is not in the document]
Comment (adambarreiro) [reply vcd/resource_vcd_network_isolated_v2.go]: Done, thanks for the suggestion!

File: vcd/resource_vcd_network_isolated_v2.go [hunk 9]
```diff
@@ -195,10 +211,20 @@ func resourceVcdNetworkIsolatedV2Read(ctx context.Context, d *schema.ResourceDat
 
 	d.SetId(orgNetwork.OpenApiOrgVdcNetwork.ID)
 
+	metadata, err := orgNetwork.GetMetadata()
+	if err != nil {
+		log.Printf("[DEBUG] Unable to find isolated network v2 metadata: %s", err)
+		return diag.Errorf("[isolated network v2 read] unable to find Org VDC network metadata %s", err)
```
[the remark this answers is not in the document]
Comment (adambarreiro) [reply vcd/resource_vcd_network_isolated_v2.go]: Fixed!

File: vcd/resource_vcd_network_isolated_v2.go [hunk 10]
```diff
@@ -195,10 +211,20 @@ func resourceVcdNetworkIsolatedV2Read(ctx context.Context, d *schema.ResourceDat
...
+	}
+	err = d.Set("metadata", getMetadataStruct(metadata.MetadataEntry))
+	if err != nil {
+		return diag.Errorf("[isolated network v2 read] unable to set Org VDC network metadata %s", err)
```
[the remark this answers is not in the document]
Comment (adambarreiro) [reply vcd/resource_vcd_network_isolated_v2.go]: Fixed!

File: vcd/resource_vcd_network_routed_v2.go (hunk 4 above)
Comment (adambarreiro) [reply vcd/resource_vcd_network_routed_v2.go]: Fixed!

**Review by vbauzysvmware — APPROVED** [review]

**Comment by adambarreiro** [comment]
Checking an issue that `TestAccVcdNetworkRoutedV2NsxtMigration` showed. It seems metadata GET is failing when the network belongs to a VDC Group. So PR can't be merged as it is.

File: .changes/v3.6.0/816-improvements.md (hunk 3 above)
Comment (lvirbalas) [reply .changes/v3.6.0/816-improvements.md]: Ah! @Didainius what do you think?

**Review by Didainius — APPROVED** [review]

**Comment by adambarreiro** [comment]
Tests passed OK

**Review by mikeletux — APPROVED** [review]

**Pull request merged**